### PR TITLE
sys-boot/raspberrypi-firmware: cleanup and update, the git purist way

### PR DESCRIPTION
**Pull request by andrey-utkin**
https://bugs.gentoo.org/671238

Original pull request by Conrad (which left me slightly unsatisfied): https://github.com/gentoo/gentoo/pull/11381

I haven't tested this on hardware yet, will do this weekend.

@ConiKost @juippis your review is very welcome.

### CODE REVIEW

**Comment by juippis** [comment]
Well, I don't really know what was wrong in @ConiKost 's ebuild, and you don't have to explain it to me since I have no interest in maintaining this :P but for the record, ebuilds look good apart from 2 `cd` commands missing `|| die`.

**Comment by andrey-utkin** [comment]
> Well, I don't really know what was wrong in @ConiKost 's ebuild, and you don't have to explain it to me since I have no interest in maintaining this :P

Just for record, his changes go in a bit too big chunks and not very well commented on what exactly is changed and why. As I was replaying basically the same changes, I notices several mistakes in his changes, which I couldn't notice from review, exacly for reasons I've described above.

> but for the record, ebuilds look good apart from 2 `cd` commands missing `|| die`.

It's legacy, but good point, thanks. Fixed.

**Comment by andrey-utkin** [comment]
Merged as c1b8b3e..8087386.

File: sys-boot/raspberrypi-firmware/raspberrypi-firmware-1.20190215.ebuild [hunk 1]
```diff
@@ -1,32 +1,27 @@
...
 LICENSE="GPL-2 raspberrypi-videocore-bin"
 SLOT="0"
-KEYWORDS="~arm -*"
-IUSE=""
-
-DEPEND=""
-RDEPEND="!sys-boot/raspberrypi-loader"
 
-S=${WORKDIR}
+if [[ "${PV}" == 9999 ]]; then
```
Comment (ConiKost) [inline sys-boot/raspberrypi-firmware/raspberrypi-firmware-1.20190215.ebuild]: What's the point in added this in a non 9999 ebuild?

File: sys-boot/raspberrypi-firmware/raspberrypi-firmware-9999.ebuild [hunk 2]
```diff
@@ -52,4 +53,8 @@ src_install() {
 	readme.gentoo_create_doc
 }
 
+pkg_postinst() {
+	readme.gentoo_print_elog
+}
+
 DOC_CONTENTS="Please configure your ram setup by editing /boot/config.txt"
```
Comment (ConiKost) [inline sys-boot/raspberrypi-firmware/raspberrypi-firmware-9999.ebuild]: That's don't needed IMO, as for default, you don't have to do this.

File: sys-boot/raspberrypi-firmware/raspberrypi-firmware-9999.ebuild [hunk 3]
```diff
@@ -42,8 +43,8 @@ pkg_preinst() {
 
 src_install() {
 	insinto /boot
-	cd boot
-	doins bootcode.bin COPYING.linux fixup*.dat LICENCE.broadcom start*elf
+	cd boot || die
+	doins bootcode.bin fixup*.dat start*elf
 	doins *.dtb
 	doins -r overlays
```
Comment (ConiKost) [inline sys-boot/raspberrypi-firmware/raspberrypi-firmware-9999.ebuild]: Do we really want overlays? If a user compiles a kernel, this installs the overlays.

File: sys-boot/raspberrypi-firmware/raspberrypi-firmware-9999.ebuild [hunk 4]
```diff
@@ -42,8 +43,8 @@ pkg_preinst() {
 
 src_install() {
 	insinto /boot
-	cd boot
-	doins bootcode.bin COPYING.linux fixup*.dat LICENCE.broadcom start*elf
+	cd boot || die
```
Comment (ConiKost) [inline sys-boot/raspberrypi-firmware/raspberrypi-firmware-9999.ebuild]: Instead of ```cd```, we could squash those doins in one single line.

**Comment by andrey-utkin** [comment]
On Sat, Mar 30, 2019 at 05:09:40AM -0700, Conrad Kostecki wrote:
> ConiKost commented on this pull request.
> 
> 
> 
> > @@ -42,8 +43,8 @@ pkg_preinst() {
>  
>  src_install() {
>  	insinto /boot
> -	cd boot
> -	doins bootcode.bin COPYING.linux fixup*.dat LICENCE.broadcom start*elf
> +	cd boot || die
> 
> Instead of ```cd```, we could squash those doins in one single line.

But it was this way before, and it worked, so I didn't spend time and
take risk by refatoring it. We can do it anytime.

My goal in this PR was to upkeep the package to make it usable again.
I was (and still am) quite constrained in time and ability to test in
runtime. So I avoided making extra changes, one of motives was basically
to avoid making a mistake in an unnecessary change.

**Comment by andrey-utkin** [comment]
On Sat, Mar 30, 2019 at 05:09:01AM -0700, Conrad Kostecki wrote:
> ConiKost commented on this pull request.
> 
> 
> 
> > @@ -42,8 +43,8 @@ pkg_preinst() {
>  
>  src_install() {
>  	insinto /boot
> -	cd boot
> -	doins bootcode.bin COPYING.linux fixup*.dat LICENCE.broadcom start*elf
> +	cd boot || die
> +	doins bootcode.bin fixup*.dat start*elf
>  	doins *.dtb
>  	doins -r overlays
> 
> Do we really want overlays? If a user compiles a kernel, this installs the overlays.

Again - you didn't tell, I didn't know, I kept it the old way.

Happy to make the change to not install overlays.

**Comment by andrey-utkin** [comment]
On Sat, Mar 30, 2019 at 05:08:35AM -0700, Conrad Kostecki wrote:
> ConiKost commented on this pull request.
> 
> 
> 
> > @@ -52,4 +53,8 @@ src_install() {
>  	readme.gentoo_create_doc
>  }
>  
> +pkg_postinst() {
> +	readme.gentoo_print_elog
> +}
> +
>  DOC_CONTENTS="Please configure your ram setup by editing /boot/config.txt"
> 
> That's don't needed IMO, as for default, you don't have to do this.

Agreed - this warning looks quite *arbitrary*, I just didn't change it
this time.
Totally happy to nuke it.

**Comment by andrey-utkin** [comment]
On Sat, Mar 30, 2019 at 05:07:08AM -0700, Conrad Kostecki wrote:
> ConiKost commented on this pull request.
> 
> 
> 
> >  
> -S=${WORKDIR}
> +if [[ "${PV}" == 9999 ]]; then
> 
> What's the point in added this in a non 9999 ebuild?

Oh it's very useful!

It is quite widespread practice to use such "unified" ebuilds.

This allows you to just copy the file to add new version, and it works
literally without any changes. Naturally, you get 9999 ebuild and new
version ebuild with exactly same content, with exactly same checksums,
and `diff` shows nothing.

So old versions get phased out as time goes, you keep editing 9999 as
your "master" ebuild, you keep cloning it into release ebuilds, and
that's it.

Benefits:
* You don't have to do all the work twice to update both release ebuild
  and 9999 ebuild.
* 9999 is never phased out so it lives forever in Gentoo tree, and `git
  blame` is really informative on it! `git blame` on other ebuilds is
  not helpful at all, all you see is when somebody added a new version.
  To get the history of some change, you need to `git log
  foo-1.2.3.ebuild` and `git checkout <that hash>` and `git log
  foo-1.2.2.ebuild` and `git checkout <that hash>` and `git log
  foo-1.2.1.ebuild` and on and on and on.

**Comment by ConiKost** [comment]
> But it was this way before, and it worked, so I didn't spend time and take risk by refatoring it. We can do it anytime. My goal in this PR was to upkeep the package to make it usable again. I was (and still am) quite constrained in time and ability to test in runtime. So I avoided making extra changes, one of motives was basically to avoid making a mistake in an unnecessary change.

Well, it's just, that i am getting from other devs the info and nit picks to make my ebuilds more compact, like this. Since I tested this, I don't see, what could got wrong.

> This allows you to just copy the file to add new version, and it works literally without any changes. Naturally, you get 9999 ebuild and new version ebuild with exactly same content, with exactly same checksums, and `diff` shows nothing.

Well okay, I see the point. If you like it this okay, fine for me. Still, IMHO, bloating the code in non live ebuilds ;-)

**Comment by andrey-utkin** [comment]
On Sun, Mar 31, 2019 at 04:09:05PM -0700, Conrad Kostecki wrote:
> > But it was this way before, and it worked, so I didn't spend time and take risk by refatoring it. We can do it anytime. My goal in this PR was to upkeep the package to make it usable again. I was (and still am) quite constrained in time and ability to test in runtime. So I avoided making extra changes, one of motives was basically to avoid making a mistake in an unnecessary change.
> 
> Well, it's just, that i am getting from other devs the info and nit picks to make my ebuilds more compact, like this. Since I tested this, I don't see, what could got wrong.

As I said, I see where from the idea of the change is coming. I protest
against the changes going without any notice :)

These changes have to be made in separate commits and be clearly
explained in commit messages. You can also refer to names of who
suggested it, using 'Suggested-by:' line in commit message. This is
really important. And it is what devmanual actually means when it says
"one logical change per commit".

It's true that this ebuild deserves an overhaul. And it's much easier to
just go through it in one pass in text editor, make it pretty, and
commit the result. It is very painful and slow to split the overhaul
into a dozen of commits, writing more text in commit messages than in
the code. But the quick way, it's just not possible for reviewers to vet
for the correctness of such change. Unless they are either genius, or
have a lot of time, or are very familiar with the code.

> 
> > This allows you to just copy the file to add new version, and it works literally without any changes. Naturally, you get 9999 ebuild and new version ebuild with exactly same content, with exactly same checksums, and `diff` shows nothing.
> 
> Well okay, I see the point. If you like it this okay, fine for me. Still, IMHO, bloating the code in non live ebuilds ;-)

We have very different values :)
I am not really concerned about extra ~4 lines of code in the ebuild.
This really saves me time and thinking efforts. I always know where is
my master ebuild which I'm going to upkeep and then clone into new
versions.

**Comment by ConiKost** [comment]
> As I said, I see where from the idea of the change is coming. I protest against the changes going without any notice :) 

Yes, I think, its fine for now, as I will split you in future the commits, as said.


> We have very different values :)

I know, that's why its fine for me, if its your current workflow :)

**Pull request closed without merging**